lib/player.js

```diff
@@ -3796,70 +3798,116 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
    * @param {string=} mimeType
    * @param {string=} codec
    * @param {string=} label
-   * @return {shaka.extern.Track}
+   * @return {!Promise.<shaka.extern.Track>}
    * @export
    */
-  addTextTrack(uri, language, kind, mimeType, codec, label) {
-    // TODO: Add an actual error for this.
+  async addTextTrack(uri, language, kind, mimeType, codec, label) {
+    if (this.loadMode_ != shaka.Player.LoadMode.MEDIA_SOURCE &&
+        this.loadMode_ != shaka.Player.LoadMode.SRC_EQUALS) {
+      shaka.log.error(
+          'Must call load() and wait for it to resolve before adding text ' +
+          'tracks.');
+      return Promise.reject(() => {
+        return new shaka.util.Error(
+            shaka.util.Error.Severity.RECOVERABLE,
+            shaka.util.Error.Category.PLAYER,
+            shaka.util.Error.Code.PLAYER_UNINITIALIZED);
+      });
+    }
+
+    if (!mimeType) {
+      // Try using the uri extension.
+      const extension = shaka.media.ManifestParser.getExtension(uri);
+      mimeType = {
+        'srt': 'text/srt',
+        'vtt': 'text/vtt',
+        'webvtt': 'text/vtt',
+        'ttml': 'application/ttml+xml',
+      }[extension];
+
+      if (!mimeType) {
+        try {
+          // eslint-disable-next-line require-atomic-updates
+          mimeType = await shaka.media.ManifestParser.getMimeType(uri,
+              this.createNetworkingEngine(),
```

Comment:
I don't believe you should use createNetworkingEngine here.  There should already be an instance, and you should use that.  Otherwise, an application's configured filters may not apply to this request.